flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.jobmaster.JobManagerRunner;
+import org.apache.flink.runtime.messages.Acknowledge;
+import org.apache.flink.runtime.rpc.RpcUtils;
+import org.apache.flink.util.function.SupplierWithException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Wrapper for an initializing JobManagerRunner.
+ */
+public class InitializingJobManagerRunner {
+	private final Logger log = LoggerFactory.getLogger(InitializingJobManagerRunner.class);
+
+	private final CompletableFuture<JobManagerRunner> jobManagerRunnerFuture = new CompletableFuture<>();
+
+	private final Thread initializationRunner;
+
+	private final Object lock = new Object();
+
+	@GuardedBy("lock")
+	private CompletableFuture<Acknowledge> cancellationFuture;
+
+	public InitializingJobManagerRunner(SupplierWithException<JobManagerRunner, JobInitializationException> initializationRunnable, JobID jobID) {
+		String threadName = "jobmanager-initialization-" + jobID;
+		this.initializationRunner = new Thread(() -> {
+			log.debug("Starting JobManager initialization thread: {}", threadName);
+			try {
+				runInitialization(initializationRunnable);
+			} catch (Throwable threadError) {
+				log.warn("Unexpected error in thread '{}'", threadName, threadError);
+			} finally {
+				log.debug("Stopping JobManager initialization thread: {}", threadName);
+			}
+		}, threadName);
+		this.initializationRunner.setDaemon(true); // this should not block the JVM from shutting down
+		this.initializationRunner.start();
+	}
+
+	private void runInitialization(SupplierWithException<JobManagerRunner, JobInitializationException> initializationRunnable) {
+		try {
+			JobManagerRunner runner = initializationRunnable.get();
+			synchronized (lock) {
+				if (cancellationFuture == null) {
+					jobManagerRunnerFuture.complete(runner);
+				} else {
+					// jobMaster has finished despite the interruption. Cancel job.
+					FutureUtils.forward(runner.getJobMasterGateway().thenCompose(gateway -> gateway.cancel(RpcUtils.INF_TIMEOUT)),
+						cancellationFuture);
+					jobManagerRunnerFuture.complete(runner);
+				}
+			}
+		} catch (JobInitializationException throwable) {
+			synchronized (lock) {
+				if (cancellationFuture == null) {
+					jobManagerRunnerFuture.completeExceptionally(throwable);
+				} else {
+					// an exception related to an interruption is expected here.
+					log.debug("Cancellation finished with (expected) exception", throwable);
+					jobManagerRunnerFuture.completeExceptionally(new JobInitializationCancelledException("Job initialization has been cancelled", throwable));
+					cancellationFuture.complete(Acknowledge.get());
+				}
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancelInitialization() {
+		synchronized (lock) {
+			initializationRunner.interrupt();
```

Comment:
The thread will also stop if it enters a (properly implemented) blocking operation.

The classical case we are targeting is that a remote system (triggered by initializeOnMaster(), or some state restore operation) is not available. The code will be in blocking operations all the time (waiting for network, sleeping, ...).
The bigger question is whether the thread is cooperative in reacting to the interrupt. Often, interrupt exceptions are ignored.

Reply:
Since we intend to move the `initializeOnMaster` execution to the IO executor in FLINK-19069, what's your plan to handle this? (do we even need FLINK-19069 after this PR?)

Reply:
FLINK-19069 is about finalizeOnMaster, not initializeOnMaster.